containers/withAdminContainer.ts

```diff
@@ -0,0 +1,34 @@
+import { Context } from '../@types/helpers'
+import { isAdmin, isAdminOrThrow } from '../helpers/isAdmin'
+import { withUserContainer } from './withUserContainer'
+import _ from 'lodash'
+
+//use when only checking if user is admin
+export const withAdminContainer =
+  <Type, ArgsType>(
+    resolver: (_parent: void, args: ArgsType, ctx: Context) => Type,
+    errorMessage?: string
+  ) =>
+  async (_parent: void, args: ArgsType, ctx: Context) => {
+    const { req } = ctx
+
+    //check if is admin
+    if (errorMessage) {
```

Comment:
Lets simplify this to:

```
if (!isAdmin(req)) {
  throw new Error(errorMessage || 'User is not an admin')
}
```

After updating all resolvers that check for `admin` with the withAdminContainer and withAdminUserContainer, we can remove the isAdminOrThrow function since the containers in this file basically replace it

Reply:
@anthonykhoa I don't think we can remove it entirely because some checks that are unique to a specific resolver can't be replicated inside a container. for example the below from updateExercise:
```
 const authorId = req.user?.id

  if (!isAdmin(req) && exercise?.authorId !== authorId)
    throw new Error('Not authorized to change')

```

Reply:
The above code is checking if this person is either an admin or the original author/creator of the exercise(in this case, they dont have to be an admin). So unless there is a way we can capture the second part of the check `exercise?.authorId !== authorId` in a container, without making it too complicated, i think we should keep as is.

if we remove the `isAdmin` here, apply the `withAdminContainer` and changed it to just:  

`if (exercise?.authorId !== authorId)
    throw new Error('Not authorized to change')`

the container will throw an error and not allow the original author(if they are not an admin) to change the exercise.

Reply:
> i think we should keep as is.

Can you explain the benefit of keeping the isAdmin function call check in this resolver? I don't understand why we are doing this check inside the resolver, since we made the withAdminContainer container so that we don't to do checks for admin status inside the resolvers anymore.

Reply:
> So unless there is a way we can capture the second part of the check exercise?.authorId !== authorId in a container,

Why not just keep this check in the updateExercise resolver itself?

Reply:
When I was working on it, and tried to apply the `withAdminContainer` to the `updateExercise` resolver, I noticed tests were failing. Below compares the behavior we ideally want, vs what happens when we apply the container, and do the check (exercise?.authorId !== authorId) by itself while wrapped in a  `withAdminContainer`:

![image](https://user-images.githubusercontent.com/77421872/186752843-f8e787cd-9007-4c3f-8f43-df42580598e5.png)

The way the resolver is set up, it should throw an error only if the user isnt an admin `&&` if the user isnt the author of the exercise. With the container, it will throw an error if the user isnt admin `||` if the user isnt the author.

Reply:
So is it just because tests are failing? When we use withAdminContainer and keep `exercise?.authorId !== authorId` inside the updateExercise resolver, it still ensures that the user must be both an admin and also the author of the exercise in order to make an update to an exercise in updateExercise resolver, right?

Reply:
Oh , i see what you are saying now

Reply:
Lets continue using the isAdmin function in this container for now then. Just do the refactoring necessary so that you don't have the `isAdmin` and `throw new Error` twice in this resolver(lines 17 and 20).